ESQL:  Replace grouping by `DateFormat` with `DateTrunc`

kanoshiou opened the pull request:
Optimize date grouping with formatting in ESQL

This PR optimizes the performance of date grouping operations in ESQL by automatically converting `DATE_FORMAT` in `GROUP BY` clauses to more efficient `DATE_TRUNC` operations. The optimization:

- Automatically detects and converts `DATE_FORMAT` patterns to equivalent `DATE_TRUNC` intervals
- Moves date formatting from the grouping phase to a subsequent EVAL phase
- Handles timezone and DST transitions correctly
- Supports various time intervals from nanoseconds to years

Example optimization:
```sql
FROM test
| STATS avg = AVG(salary) BY date = DATE_FORMAT("yyyy-MM", hire_date)
```
becomes:
```sql
FROM test
| STATS avg = AVG(salary) BY date1 = DATE_TRUNC(1 month, hire_date) 
| EVAL date = DATE_FORMAT("yyyy-MM", date1) 
| KEEP avg, date
```
Optimized plan:
```
Project[[avg{r}#7, date{r}#4]]
\_Eval[[$$SUM$avg$0{r$}#21 / $$COUNT$avg$1{r$}#22 AS avg#7, DATEFORMAT([79 79 79 79 2d 4d 4d][KEYWORD],$$DATE_FORMAT(
"yy>$date$0{r$}#20) AS date#4]]
  \_Limit[1000[INTEGER],false]
    \_Aggregate[[$$DATE_FORMAT("yy>$date$0{r$}#20],[SUM(salary{f}#14,true[BOOLEAN]) AS $$SUM$avg$0#21, COUNT(salary{f}#14,true[
BOOLEAN]) AS $$COUNT$avg$1#22, $$DATE_FORMAT("yy>$date$0{r$}#20]]
      \_Eval[[DATETRUNC(P1M[DATE_PERIOD],hire_date{f}#16) AS $$DATE_FORMAT("yy>$date$0#20]]
        \_EsRelation[test][_meta_field{f}#15, emp_no{f}#9, first_name{f}#10, g..]
```


Closes #114772

kanoshiou committed:
Replace grouping by `DateFormat` With `DateTrunc`

kanoshiou committed:
Replace in `ReplaceAggregateNestedExpressionWithEval`

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

kanoshiou committed:
Avoid incorrect grouping with multiple DATE_FORMAT

kanoshiou committed:
Update docs/changelog/129277.yaml

kanoshiou committed:
Update csv test

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

kanoshiou committed:
Merge branch 'main' into optimize-date-grouping-with-formatting

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregateNestedExpressionWithEval.java
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

kanoshiou committed:
precommit

kanoshiou committed:
ES|QL: No plain strings in Literal #129399

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregateNestedExpressionWithEval.java:
```diff
@@ -193,4 +276,43 @@ private static Expression transformAggregateFunction(
         return TemporaryNameUtils.temporaryName(expression, func, counter);
     }
+
+    private static Literal formatToMinimalInterval(String format, Source source) {
```
It looks like we need more tests here to complete the coverage of the varieties of format pattern, to validate the conversions from the format string to intervals are correct.

fang-xing-esql reviewed (commented):
Thank you for picking this up, @kanoshiou! This is not a trivial feature, so it will likely require multiple iterations. `ReplaceAggregateNestedExpressionWithEval` seems like a good place for this rewrite.

I left some comments in the rule and the test. One challenge is that we don’t currently have many tests where `date_format` is referenced by `Aggregate`. If we move forward with this rewrite, it would be a good opportunity to expand test coverage and exercise as many code paths as possible. Since aggregate is a key command supporting stats, inlinestats, and time-series queries, having comprehensive  test coverage here would give us more confidence in the change.

fang-xing-esql commented:
buildkite test this

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/date.csv-spec

kanoshiou commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:
```diff
@@ -7860,6 +7862,85 @@ public void testSampleNoPushDownChangePoint() {
+     *   \_Limit[1000[INTEGER],false]
+     *     \_Aggregate[[date{r}#4],[SUM(salary{f}#14,true[BOOLEAN]) AS $$SUM$avg$0#20, COUNT(salary{f}#14,true[BOOLEAN]) AS $$COUNT$av
+     * g$1#21, date{r}#4]]
+     *       \_Eval[[DATEFORMAT([79 79 79 79][KEYWORD],hire_date{f}#16) AS date#4]]
```
You’re right. I must have pasted the wrong plan by mistake. I committed this code in the middle of the night, so my brain was probably in full chaos mode 😅

kanoshiou committed:
Review

kanoshiou committed:
Merge branch 'main' into optimize-date-grouping-with-formatting

kanoshiou committed:
Reuse alias name

fang-xing-esql commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregateNestedExpressionWithEval.java:
```diff
@@ -61,6 +106,32 @@ protected LogicalPlan rule(Aggregate aggregate) {
                     // Move the alias into an eval and replace it with its attribute.
                     groupingChanged = true;
                     var attr = as.toAttribute();
+                    if (asChild instanceof DateFormat df && dateFormatCount[0] == 1) {
```
Thanks for the explanation, it makes sense. After giving it more thought, I realized this issue isn’t limited to multiple `date_format` usages. More generally, we need to be careful about how patterns map to intervals to ensure consistent semantics.

Using the date_format("dd", @timestamp) as an example, this query returns different results between main and this branch.
```
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from sample_data | eval x= date_format(\"dd\", @timestamp)"
}
'
       @timestamp       |   client.ip   |event.duration |       message       |      mv       |       test_date        |       x       
------------------------+---------------+---------------+---------------------+---------------+------------------------+---------------
2022-10-23T12:15:03.360Z|172.21.2.162   |3450233        |Connected to 10.1.0.3|[1, 2]         |1985-11-21T00:00:00.000Z|23             
2023-09-23T12:27:28.948Z|172.21.2.113   |2764889        |Connected to 10.1.0.2|[1, 2]         |1985-11-22T00:00:00.000Z|23             
2020-10-23T13:33:34.937Z|172.21.0.5     |1232382        |Disconnected         |[1, 2]         |1985-11-23T00:00:00.000Z|23             
2023-08-23T13:51:54.732Z|172.21.3.15    |725448         |Connection error     |[1, 2]         |1985-11-24T00:00:00.000Z|23             
2024-10-23T13:52:55.015Z|172.21.3.15    |8268153        |Connection error     |[1, 2]         |1985-11-25T00:00:00.000Z|23             
2023-07-23T13:53:55.832Z|172.21.3.15    |5033755        |Connection error     |[1, 2]         |1985-11-26T00:00:00.000Z|23             
2023-10-23T13:55:01.543Z|172.21.3.15    |1756467        |Connected to 10.1.0.1|[1, 2]         |1985-11-27T00:00:00.000Z|23             

**on the branch**
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from sample_data | stats count(*) by date_format(\"dd\", @timestamp)"
}
'
   count(*)    |date_format("dd", @timestamp)
---------------+-----------------------------
1              |23                           
1              |23                           
1              |23                           
1              |23                           
1              |23                           
1              |23                           
1              |23   

**on main**
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from sample_data | stats count(*) by date_format(\"dd\", @timestamp)"
}
'
   count(*)    |date_format("dd", @timestamp)
---------------+-----------------------------
7              |23     
                        
```

kanoshiou replied:
Thanks for your information! I admittedly overlooked the issue with 'dd'. Given that, I believe we need to verify whether the format meets the following criteria: it must contain a sequence of progressively smaller time units starting from 'year' (with 'week' and 'quarter' considered equivalent to 'month'). If this condition is satisfied, then even multiple `DATE_FORMAT` should be eligible for optimization. However, if the format does not meet this requirement, we should avoid optimizing `DATE_FORMAT` using `DATE_TRUNC`.

kanoshiou committed:
Refine format-to-trunc interval inference logic

kanoshiou committed:
Fix continuous time unit check

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/ReplaceAggregateNestedExpressionWithEval.java

kanoshiou committed:
DateFormatToTruncIntervalTests

kanoshiou committed:
Update ReplaceAggregateNestedExpressionWithEval

kanoshiou committed:
Support `QUARTER_OF_YEAR`

kanoshiou committed:
Update stats.csv-spec

kanoshiou committed:
Support for inline stats

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

kanoshiou committed:
Precommit

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

kanoshiou commented:
Thanks for your review, @fang-xing-esql! I’ve updated the branch based on your comments.

I revised the logic for inferring the minimal time interval corresponding to a given date format. Note that `date_nanos` cannot be optimized, as nanosecond intervals are not supported by `DATE_TRUNC`. I also updated `LogicalPlanOptimizerTests` so that this optimization supports inline stats, and added additional tests covering the entire change.

I’d appreciate it if you could take another look when you have time.

fang-xing-esql reviewed (commented):
Thank you for adding more tests and handle inlinestats @kanoshiou! I added some comment around the casting of the format to `literal`.

I'll need to look into the change related to `inline stats` more, as it seems a bit more complicated than I had expected. This change created an `eval` on top of `aggregate`, which makes `inline stats` a bit tricky.

And I'm wondering if you could elaborate a bit more about why this transformation does not apply to `date_nanos`? I tried some queries on `date_nanos`, and they seem work fine, the changes to the rule doesn't block `date_nanos`. Please find my experiments below(the 2nd and 4th query show equivalent results). Perhaps there is something that I missed here.

```
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from test1 | eval x= date_format(\"y-MM-dd\", nanos)"
}
'
         millis         |                                            nanos                                            |        num        |       x       
------------------------+---------------------------------------------------------------------------------------------+-------------------+---------------
2023-10-23T13:55:01.543Z|2023-10-23T13:55:01.543123456Z                                                               |1698069301543123456|2023-10-23     
2023-10-23T13:55:01.543Z|2023-10-23T12:55:01.543123456Z                                                               |1698069301543123456|2023-10-23     
1999-10-23T12:15:03.360Z|[2023-01-23T13:55:01.543123456Z, 2023-02-23T13:33:34.937193Z, 2023-03-23T12:15:03.360103847Z]|0                  |null           
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from test1 | stats count(*) by date_format(\"y-MM-dd\", nanos)"
}
'
   count(*)    |date_format("y-MM-dd", nanos)
---------------+-----------------------------
1              |null                         
2              |2023-10-23                                     
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from test1 | eval x= date_trunc(1 day, nanos), y= date_format(\"y-MM-dd\", x)"
}
'
         millis         |                                            nanos                                            |        num        |           x            |       y       
------------------------+---------------------------------------------------------------------------------------------+-------------------+------------------------+---------------
2023-10-23T13:55:01.543Z|2023-10-23T13:55:01.543123456Z                                                               |1698069301543123456|2023-10-23T00:00:00.000Z|2023-10-23     
2023-10-23T13:55:01.543Z|2023-10-23T12:55:01.543123456Z                                                               |1698069301543123456|2023-10-23T00:00:00.000Z|2023-10-23     
1999-10-23T12:15:03.360Z|[2023-01-23T13:55:01.543123456Z, 2023-02-23T13:33:34.937193Z, 2023-03-23T12:15:03.360103847Z]|0                  |null                    |null           
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from test1 | stats count(*) by x = date_trunc(1 day, nanos) | eval y= date_format(\"y-MM-dd\", x)"
}
'
   count(*)    |           x            |       y       
---------------+------------------------+---------------
1              |null                    |null           
2              |2023-10-23T00:00:00.000Z|2023-10-23  
```

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/inlinestats.csv-spec
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PropagateInlineEvals.java

kanoshiou committed:
Merge remote-tracking branch 'origin/main' into optimize-date-grouping-with-formatting

kanoshiou committed:
Merge remote-tracking branch 'origin/main' into optimize-date-grouping-with-formatting

kanoshiou committed:
Fix unsafe cast of date format when not a constant

kanoshiou committed:
Add more tests

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

kanoshiou commented:
Thanks for your review @fang-xing-esql!

I thought you were referring to formatting nanosecond intervals using `date_format`. This transformation also applies to the `date_nanos` type, and I’ve updated the branch and added additional tests to cover both the `date_nanos` case and scenarios where the format is a `ReferenceAttribute`.

fang-xing-esql reviewed (commented):
Thank you for addressing the previous round of comments and for being patient, @kanoshiou! I went through the code changes again, and for most queries, the PR does a great job transforming `date_format` to `date_trunc` correctly — really appreciate the effort.

I’ve added a few comments in the code suggesting minor updates to the tests and additional explanations in areas that are a bit tricky to follow. Since aggregation can get complex and tricky, having broader test coverage will help ensure that future changes don’t unintentionally alter expected behavior.

I have tried some normal queries, they work as expected. And I'm trying to push it harder with a bit more tricky queries, and came across these situations. In `ES|QL` , we allow the flexibility that `grouping` can be referenced in the `aggregate` explicitly, and sometime this flexibility adds extra complexity. These two queries below don't error out in main.

```
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from test1 | stats count(*), concat(x, \"01\") by x = date_format(\"y-MM-dd\", nanos)"
}
'
{
  "error" : {
    "root_cause" : [
      {
        "type" : "null_pointer_exception",
        "reason" : "Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.esql.core.expression.NameId)\" is null"
      }
    ],
    "type" : "null_pointer_exception",
    "reason" : "Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.esql.core.expression.NameId)\" is null"
  },
  "status" : 500
}
+ curl -u elastic:password -X POST 'localhost:9200/_query?format=txt&pretty' -H 'Content-Type: application/json' '-d
{
  "query": "from test1 | inline stats count(*), concat(x, \"01\") by x = date_format(\"y-MM-dd\", nanos)"
}
'
{
  "error" : {
    "root_cause" : [
      {
        "type" : "illegal_state_exception",
        "reason" : "Found 1 problem\nline 1:14: Plan [Eval[[CONCAT(x{r}#106,01[KEYWORD]) AS concat(x, \"01\")#103]]] optimized incorrectly due to missing references [x{r}#106]"
      }
    ],
    "type" : "illegal_state_exception",
    "reason" : "Found 1 problem\nline 1:14: Plan [Eval[[CONCAT(x{r}#106,01[KEYWORD]) AS concat(x, \"01\")#103]]] optimized incorrectly due to missing references [x{r}#106]"
  },
  "status" : 500
}
```

kanoshiou committed:
Update tests

kanoshiou committed:
Update `PropagateInlineEvals`

kanoshiou commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PropagateInlineEvals.java:
```diff
@@ -86,6 +89,57 @@ protected LogicalPlan rule(InlineJoin plan) {
+
+                for (Alias f : eval.fields()) {
+                    if (f.child() instanceof DateFormat df && shadowing.remove(df.field())) {
+                        Attribute original = (Attribute) df.field();
```
Calling `shadowing.remove(df.field())` ensures that `df.field()` is at least a `NamedExpression`. I’ll update this part to make it clearer.

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/inlinestats.csv-spec
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

kanoshiou committed:
Update

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

kanoshiou commented:
I sincerely appreciate the time and effort you put into reviewing this, @fang-xing-esql! Thanks for your review! Your edge cases helped me deepen my understanding of `ES|QL`. I've updated the branch based on your comments - thanks again for the insightful feedback, and I’d appreciate it if you could take another look when you have time.

Regarding the case where the `grouping` is referenced within the `aggregate`, I’ve given it some thought. In such a scenario, it seems that the optimization cannot be applied, since the `aggregate` operation requires the formatted field to be present upfront. What do you think? I’d appreciate any feedback or thoughts you might have on this.

kanoshiou committed:
Clean code

kanoshiou committed:
Merge branch 'main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec

fang-xing-esql committed:
Merge branch 'main' into optimize-date-grouping-with-formatting

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/inlinestats.csv-spec

kanoshiou commented:
Thanks for the detailed feedback and I completely understand your concerns, @fang-xing-esql.

I'm considering adding some benchmarks for this PR. However, after reviewing the existing code under the `benchmark` package, I couldn’t find any examples that directly apply to full `ES|QL` query execution. From what I can tell, it might be necessary to introduce a new benchmark specifically designed to run complete `ES|QL` queries end-to-end. Does that sound reasonable to you?

As for timezone handling, I believe this PR should be fine as long as both `date_trunc` and `date_format` support timezone correctly.

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/inlinestats.csv-spec

kanoshiou committed:
Merge branch 'refs/heads/main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

kanoshiou committed:
Merge branch 'main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/qa/testFixtures/src/main/resources/inlinestats.csv-spec

kanoshiou committed:
Merge branch 'main' into optimize-date-grouping-with-formatting

# Conflicts:
#	x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PropagateInlineEvals.java

wchaparro commented:
<!-- analytics-pr-review-management:stall-warn -->

Hi @kanoshiou — this PR has been identified as **stalled** (open and without meaningful updates for 30+ days).

@astefan — a review (or a note if you cannot take this) would help.

It will be **closed in two weeks** if there is no further activity. Please reply, push an update, or close it yourself if it is no longer needed.

— Analytics PR review management

kanoshiou committed:
Merge branch 'main' into optimize-date-grouping-with-formatting
